### DISCUSSION TITLE
Windows 7 madness

### DISCUSSION THREAD

**Discussion by ghost** [terminatorul/NvStrapsReBar#27] [General]
Using NvStrapsRebar v0.2 prebuilts

Motherboard: Asrock Z77 Pro 3
BIOS Version: 1.91
GPU: GTX 1660 Super 6GiB
CSM: Turned **on**. W7 does not continue pass flying orbs startup before desktop if disabled.
Above 4G decoding: On 
Partitioning: GPT
UEFIPatch: For above 2GiB BAR the Ivy >4GB patch was used.

Note that W7 doesn't support get and set of UEFI variables therefore a workaround using W10Pe x64 USB boot plus VC 2015 install was used to run NvStrapsRebar and setup required BIOS variables.

CSM
![csm](https://github.com/terminatorul/NvStrapsReBar/assets/147298897/b3377596-628b-4267-aad5-ed8be4d2dcb3)

Option E provided a 2GiB BAR with no problems. Setting higher BAR sizes was achieved by using Ivy >4GB UEFIPatch.

8GiB BAR1. Discovering CSM on W7 seems problematic, possibly why HWiNFO doesn't mention it. 
![8G_Rebar](https://github.com/terminatorul/NvStrapsReBar/assets/147298897/7b32c580-cf81-4b0e-bc1f-8f70fcacc032)

Surprisingly resume doesn't crash as 8GiB BAR becomes 256MiB.
![Resume_Rebar](https://github.com/terminatorul/NvStrapsReBar/assets/147298897/3d796a79-76ae-4083-9121-78cdc833fc84)
Seems some stale data is used, not reacquired on resuming. Also GPU-Z now reporting Rebar enabled with 256MiB, perhaps using address range rather than actual size, idk. 

This raises the question of whether the W7 nvidia driver is limiting the 8GiB BAR1 to 256MiB even though nvidiaprofile was set to global resizable BAR at 1GiB and some graphics applications were run?

**Comment by xCuri0** [comment]
@Xelafic it's probably just that windows 7 handles unexpected BAR downsizing better, same is happening on windows 10 and it bsods.

And btw there are methods to get windows 7 working with CSM.

**Reply by ghost** [reply]
Hi @xCuri0 :wave: 
Do you mean 'without CSM'? Did see some mention of using bcdedit or registry in that case. I have never had a problem if CSM is left enabled although I am very limited with hardware used.

**Reply by xCuri0** [reply]
@Xelafic yes without csm is what I meant

I think a recent update from around a year ago by Microsoft allows it and there are some previous unofficial methods too.

Atleast on my gigabyte b75 I can't open the bios menu with csm + 4g decoding and also boot logo is low resolution

**Reply by ghost** [reply]
>Atleast on my gigabyte b75 I can't open the bios menu with csm + 4g decoding and also boot logo is low resolution

That sucks. Was CSM on with graphics Option ROM set for UEFI and GOP the latest? Saw some other posts with getting into BIOS setup a problem. What was bizarre is https://github.com/xCuri0/ReBarUEFI/issues/111 having problems. That's a premium Asrock Z77 that costs over 3 times the price of my 'budget' Asrock Z77 Pro3. Well I suppose for most people entering BIOS is a rare thing so perhaps some workaround disabling above 4G setting from OS and rebooting to UEFI might work well, 4G can be re-enabled at that time.

**Reply by xCuri0** [reply]
I just keep CSM off, everything works proper BIOS menus and full resolution.

**Comment by ghost** [comment]
@terminatorul tried 16GiB and 32GiB BAR1 with no luck. UEFI patches gave no post or no video, haven't time to investigate but where I was able to use a second graphics card then it showed NvStrapsRebar had done it's job well but the mainboard BIOS itself wasn't up to it.

Reverting back to running an EFI to set and reallocate after POST (no BIOS mod or patches) and chain loading Windows 32GiB BAR1 works fine as far as I can tell. No resume from sleep of course. Above 4G was disabled in the BIOS to allow use of maximum MMIO above 4G. Would have liked to have tried 32GiB of RAM but only have 24GiB at hand. It would have needed some extra tweaks to restrict top of memory and allow some small MMIO directly below 32GiB address. While CSM was enabled, the UEFI OROM was selected.
![32GiB_BAR1](https://github.com/terminatorul/NvStrapsReBar/assets/147298897/aff13ee0-f673-4871-beba-fceb832e288b)
Using W10 as my W7 home version is only licensed for 16GiB max.
I have a backlog of things to catch up on so am going to cut loose. Pleasure to have e-met, all the best for the future.

**Reply by terminatorul** [reply]
Does it really work to reallocate after POST ?

Do you have some code you could post that does that please ?
